### PR TITLE
MultiplePassReadWalker

### PR DESCRIPTION
**Pull request by tedsharpe**

### CODE REVIEW

**Comment by cmnbroad** [comment]
@tedsharpe Do you anticipate needing more than two passes ? If not, you can use `TwoPassReadWalker`.

**Comment by tedsharpe** [comment]
@cmnbroad It's not clear that I will at this point, but the SV Spark tool takes multiple passes, and what I'm goofing around with right now will be a part of that (or son of that).
I just thought it might be helpful to have this alternative available.  It's not worth spending a lot of time on.
What's nice is that the engine puts you in charge, for once.  You get to make any number of traversals if, as, and when you need them.  It relieved me of the necessity to stuff my object with a bunch of transient state.  And the ReadDataSource can be managed by a try with resources, so that looks a lot more bullet-proof than the current design, too.  (For example, the TwoPassReadWalker leaks the first ReadDataSource when it reinitializes for the second pass.)

**Comment by cmnbroad** [comment]
Well, I definitely like the safety and generality of this design better, thought it stands it does waste the first turn of the (`ReadsDataSource`) crank, which would be nice to avoid. Let me take a closer look. Maybe we can replace `TwoPassReadWalker` with this.

**Comment by tedsharpe** [comment]
One suspects that there's a lot of effort wasted by this approach.  If we dissected the ReadDataSource, we could probably figure out a way to separate the parts that need to be reinitialized for a new pass (iterators) from the parts that don't (cached index, e.g.).

File: src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java [hunk 1]
```diff
@@ -443,6 +443,10 @@ else if (hasCramInput()) {
                 factory = factory.enable(SamReaderFactory.Option.CACHE_FILE_BASED_INDEXES);
             }
 
+            if ( reads != null ) {
+                reads.close();
+            }
```
Comment (cmnbroad) [inline src/main/java/org/broadinstitute/hellbender/engine/GATKTool.java]: Yeah, if we keep TwoPassReadWalker that would be the fix, but I'd still like to consider replacing it with `MultiPassReadWalker`. I will spend some time on the latter in the next week or so, under the assumption that this isn't blocking you and is not a high priority. Do let me know if that changes.

**Comment by droazen** [comment]
@tedsharpe @cmnbroad Should we try to get this merged for the upcoming release?

**Comment by tedsharpe** [comment]
I don’t feel that there’s any great time pressure, but I do think I’ll need it at some point. And these things do have a way of withering with time...
Maybe not essential for this release.

**Comment by droazen** [comment]
Have you addressed all of @cmnbroad 's comments? If you think this branch is good to go, we might be able to get it in for this release.

**Comment by cmnbroad** [comment]
We should hold off - there is a fair amount of code that was added since I looked at the original PR, most of which hasn't been reviewed yet, and there are no tests.  I will get to the engine part this week.

**Comment by cmnbroad** [comment]
@tedsharpe Not sure what the status of this is. Is this local assembler tool something you want to put into GATK ? If so, can you rebase this on top of the new MultiplePassReadWalker base and resolve the conflicts ? thx.

**Comment by tedsharpe** [comment]
Pretty sure this is now a dead branch. I think this was my initial attempt at a multi-pass walker that you improved. My local assembler is now on a different branch, rebased on your work. It’s still a work in progress.
I’m on vacation until 1/15, and will clean this up when I return.

**Pull request closed without merging**